### PR TITLE
Test categorical features with column-split gpu quantile

### PR DESCRIPTION
**Pull request by rongou**

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

**Comment by trivialfis** [comment]
I will submit a PR for fixing the macos test tomorrow.

File: src/common/quantile.cu [hunk 1]
```diff
@@ -657,7 +657,9 @@ void SketchContainer::MakeCuts(HistogramCuts* p_cuts, bool is_column_split) {
     size_t column_size = std::max(static_cast<size_t>(1ul), this->Column(i).size());
     if (IsCat(h_feature_types, i)) {
       // column_size is the number of unique values in that feature.
-      CheckMaxCat(max_values[i].value, column_size);
+      if (!is_column_split) {
```
Comment (trivialfis) [inline src/common/quantile.cu:660]: why not?

Comment (rongou) [reply src/common/quantile.cu:660]: Hmm there seems to be a bug dealing with missing columns. Trying to fix it.

Comment (rongou) [reply src/common/quantile.cu]: Fixed the underlying issue. Please take another look.

File: src/common/quantile.cu [hunk 2]
```diff
@@ -634,12 +634,25 @@ void SketchContainer::MakeCuts(HistogramCuts* p_cuts, bool is_column_split) {
...
+    auto new_end = thrust::reduce_by_key(
+        thrust::cuda::par(alloc), key_it, key_it + in_cut_values.size(), val_it, d_max_keys.begin(),
+        d_max_values.begin(), thrust::equal_to<bst_feature_t>{},
+        [] __device__(auto l, auto r) { return l.value > r.value ? l : r; });
+    d_max_keys.erase(new_end.first, d_max_keys.end());
+    d_max_values.erase(new_end.second, d_max_values.end());
+
+    // The device vector needs to be initialized explicitly since we may have some missing columns.
+    SketchEntry default_entry{};
+    dh::caching_device_vector<SketchEntry> d_max_results(d_in_columns_ptr.size() - 1,
```
Comment (trivialfis) [inline src/common/quantile.cu:651]: Are you sure the caching device vector does initialize the value? (call constructor)

File: src/common/quantile.cu [hunk 3]
```diff
@@ -634,12 +634,25 @@ void SketchContainer::MakeCuts(HistogramCuts* p_cuts, bool is_column_split) {
...
-    thrust::reduce_by_key(thrust::cuda::par(alloc), key_it, key_it + in_cut_values.size(), val_it,
-                          thrust::make_discard_iterator(), d_max_values.begin(),
-                          thrust::equal_to<bst_feature_t>{},
-                          [] __device__(auto l, auto r) { return l.value > r.value ? l : r; });
-    dh::CopyDeviceSpanToVector(&max_values, dh::ToSpan(d_max_values));
+    auto new_end = thrust::reduce_by_key(
+        thrust::cuda::par(alloc), key_it, key_it + in_cut_values.size(), val_it, d_max_keys.begin(),
+        d_max_values.begin(), thrust::equal_to<bst_feature_t>{},
+        [] __device__(auto l, auto r) { return l.value > r.value ? l : r; });
+    d_max_keys.erase(new_end.first, d_max_keys.end());
```
Comment (trivialfis) [inline src/common/quantile.cu:646]: I'm a bit confused by these two erases, what are they doing?

Comment (rongou) [reply src/common/quantile.cu:646]: Shrink the two vectors to actual size. If we have missing columns, they won't be fully populated.

File: src/common/quantile.cu (hunk 2 above)
Comment (rongou) [reply src/common/quantile.cu:651]: Yes: https://thrust.github.io/doc/classthrust_1_1device__vector_a881860f3a5d6236bcf96272c25063bbc.html#a881860f3a5d6236bcf96272c25063bbc

Also verified it in debugger.

Comment (trivialfis) [reply src/common/quantile.cu:651]: Got it, I think we ran into trouble with it before as commented in the `XGBCcachingDeviceAllocatorImpl`. But you are correct.

**Review by trivialfis — APPROVED** [review]

File: src/common/quantile.cu [hunk 4]
```diff
@@ -634,12 +634,25 @@ void SketchContainer::MakeCuts(HistogramCuts* p_cuts, bool is_column_split) {
...
+    d_max_values.erase(new_end.second, d_max_values.end());
+
+    // The device vector needs to be initialized explicitly since we may have some missing columns.
+    SketchEntry default_entry{};
+    dh::caching_device_vector<SketchEntry> d_max_results(d_in_columns_ptr.size() - 1,
+                                                         default_entry);
+    thrust::scatter(d_max_values.begin(), d_max_values.end(), d_max_keys.begin(),
```
Comment (trivialfis) [inline src/common/quantile.cu:653]: exec policy?

Comment (rongou) [reply src/common/quantile.cu]: Done.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**